cmd/kubeadm/app/phases/upgrade/staticpods_test.go

```diff
@@ -746,3 +844,335 @@ func TestRenewCerts(t *testing.T) {
+			expectedError: true,
+		},
+		{
+			description: "desiredEtcdVersion lessThan currentEtcdVersion error",
+			cfg: &kubeadmapi.InitConfiguration{
+				LocalAPIEndpoint: kubeadmapi.APIEndpoint{AdvertiseAddress: "192.168.2.2"},
+				ClusterConfiguration: kubeadmapi.ClusterConfiguration{
+					KubernetesVersion: "1.13.1",
+				},
+			},
+			oldEtcdClient: fakeTLSEtcdClient{
+				version: "3.3.10",
+			},
+			expectedFatal: false,
+			expectedError: true,
+		},
+		{
+			description: "desiredEtcdVersion is the same as currentEtcdVersion",
+			cfg: &kubeadmapi.InitConfiguration{
+				LocalAPIEndpoint: kubeadmapi.APIEndpoint{AdvertiseAddress: "192.168.2.2"},
+				ClusterConfiguration: kubeadmapi.ClusterConfiguration{
+					KubernetesVersion: "1.14.0",
+				},
+			},
+			oldEtcdClient: fakeTLSEtcdClient{
+				version: "3.3.10",
+			},
+			expectedFatal: false,
+			expectedError: false,
+		},
+		{
+			description: "etcdVersion compare success",
+			cfg: &kubeadmapi.InitConfiguration{
+				LocalAPIEndpoint: kubeadmapi.APIEndpoint{AdvertiseAddress: "192.168.2.2"},
+				ClusterConfiguration: kubeadmapi.ClusterConfiguration{
+					KubernetesVersion: "1.14.0",
+				},
+			},
+			oldEtcdClient: fakeTLSEtcdClient{
+				version: "3.2.24",
+			},
+			expectedFatal: true,
+			expectedError: false,
+		},
+	}
+
+	for _, rt := range tests {
+		t.Run(rt.description, func(t *testing.T) {
+			t.Logf("TestCompareEtcdVersion: %s\n", rt.description)
+			actualFatal, actualError := compareEtcdVersion(rt.cfg, rt.oldEtcdClient)
+			if actualFatal != rt.expectedFatal {
+				t.Errorf("%s unexpected failure: %v", rt.description, actualFatal)
+				return
+			}
+			if (actualError != nil) && !rt.expectedError {
+				t.Errorf("%s unexpected failure: %v", rt.description, actualError)
+				return
+			} else if (actualError == nil) && rt.expectedError {
+				t.Errorf("%s passed when expected to fail", rt.description)
+				return
+			}
+		})
+	}
+}
+
+func TestRollbackEtcdOnFailedUpgrade(t *testing.T) {
+	tests := []struct {
+		description      string
+		oldEtcdClient    etcdutil.ClusterInterrogator
+		recoverManifests map[string]string
+		expectedErr      bool
+	}{
+		{
+			description:   "roll back etcd success with no recoverManifests",
+			oldEtcdClient: fakeTLSEtcdClient{},
+			expectedErr:   false,
+		},
+		{
+			description: "roll back etcd fail with no recoverManifests",
+			oldEtcdClient: fakeTLSEtcdClient{
+				availableErr: true,
+			},
+			expectedErr: true,
+		},
+		{
+			description:      "roll back etcd success with recoverManifests",
+			oldEtcdClient:    fakeTLSEtcdClient{},
+			recoverManifests: map[string]string{},
+			expectedErr:      false,
+		},
+		{
+			description: "roll back etcd fail with recoverManifests",
+			oldEtcdClient: fakeTLSEtcdClient{
+				availableErr: true,
+			},
+			recoverManifests: map[string]string{},
+			expectedErr:      true,
+		},
+	}
+
+	for _, rt := range tests {
+		t.Logf("TestRollbackEtcdOnFailedUpgrade: %s\n", rt.description)
```

Comment:
No need for the verbosity here, ideally we should be using `t.Run` and have nothing in common between the tests (so that they execute in parallel and not be flaky).

Reply:
Thanks for your review, I have modified as your suggestion.